### PR TITLE
1599 - Re-code technicians in ASC-WDS dataset

### PR DESCRIPTION
**Pull request by AchyutRatkanthiwar**
## Description
Trello ticket [#1599](https://trello.com/c/YXycm5Id/1599-re-code-technicians-in-asc-wds-datasets)

Renamed replace_care_navigator_with_care_coordinator function to backdate_job_role_changes for clarity and optimised implementation to use dictionary-based replacements targeting the main_job_role_clean column and updated corresponding test data names. Added Technician Job Role('22') in the dictionary to replace it with other_non_care_related_staff ('27').

## Testing
- [x] Unit tests passing
- [x] Successful [Step Function run](https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/v2/executions/details/arn:aws:states:eu-west-2:856699698263:execution:1599-recode-jr-Transform-ASCWDS-Data:01b77881-bb13-4e16-adc4-be278021b313)
- [x] Outputs checked in Athena - output attached to the ticket

## Checklist (delete if not relevant)
- [x] Unit tests added/amended
- [x] Docstrings added/updated
- [x] Updated CHANGELOG
- [x] Code reviewed by AI
- [x] Moved Trello ticket to PR column

## Reviewer checklist
- [x] PR solves the Trello ticket requirements
- [x] Outputs appear reasonable and understandable
- [x] The overall approach is appropriate and not over-engineered
- [x] No obvious scalability, performance or interdependency concerns
- [x] Tests appropriately cover the main behaviour/edge cases
- [x] Naming and structure are broadly understandable and consistent
- [x] Docstrings are sufficient for future maintenance
- [x] Docstrings cover non-obvious behaviour

### CODE REVIEW

File: projects/_01_ingest/ascwds/jobs/clean_ascwds_worker_data.py [hunk 1]
```diff
@@ -113,23 +113,26 @@ def create_clean_main_job_role_column(df: DataFrame) -> DataFrame:
     )
 
 
-def replace_care_navigator_with_care_coordinator(df: DataFrame) -> DataFrame:
+def backdate_job_role_changes(df: DataFrame) -> DataFrame:
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/ascwds/jobs/clean_ascwds_worker_data.py]: I think call this something more obvious.

remap_mainjrid_codes

File: projects/_01_ingest/ascwds/jobs/clean_ascwds_worker_data.py [hunk 2]
```diff
@@ -113,23 +113,26 @@ def create_clean_main_job_role_column(df: DataFrame) -> DataFrame:
...
+    job role column and 'Technician' ("22") with
+    'Other non-care Related Staff' ("27") in the main job role column.
 
     In May 2024, the job role 'Care Navigator' was removed from ASC-WDS and all
     workers in ASC-WDS in that role at the time were moved to the 'Care
-    Co-ordinator' role. This function backdates this change to the start of the
-    dataset for consistency.
+    Co-ordinator' role. Similar thing happened with 'technician' ("22") which was
+    moved to 'Other non-care Related Staff' ("27"). This function backdates
+    these change to the start of the dataset for consistency.
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/ascwds/jobs/clean_ascwds_worker_data.py]: Both these job role changes happened at the same time. Could make this easier to update in the future by saying what the function does, then bullet points to show role and date it was removed from ASC-WDS.

"This function recodes job roles by replacing raw mainjrid codes. 

Over time, some job roles have stopped being collected by the ASC-WDS. Worker records with these roles were recoded into different roles. For consistency across all periods, this function applies that mapping across all periods.
The notes below show the date at which roles were removed from the ASC-WDS:
    - May 2024: 'Care navigator', 'Technician'.
"

File: projects/_01_ingest/unittest_data/ingest_test_file_data.py [hunk 3]
```diff
@@ -301,17 +301,19 @@ class ASCWDSWorkerData:
         ("141", date(2025, 1, 1), "41", "40", MainJobRoleLabels.care_coordinator),
     ]
 
-    replace_care_navigator_with_care_coordinator_values_updated_when_care_navigator_is_present_rows = [
+    backdate_job_role_changes_mapped_codes_rows = [
         ("41", "41"),
+        ("22", "22"),
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/unittest_data/ingest_test_file_data.py]: First column in this is workerid, make the value four digits so it doesn't look like a job role code.

File: projects/_01_ingest/unittest_data/ingest_test_file_data.py [hunk 4]
```diff
@@ -301,17 +301,19 @@ class ASCWDSWorkerData:
...
-    expected_replace_care_navigator_with_care_coordinator_values_updated_when_care_navigator_is_present_rows = [
+    expected_backdate_job_role_changes_mapped_codes_rows = [
         ("41", "40"),
+        ("22", "27"),
     ]
-    replace_care_navigator_with_care_coordinator_values_remain_unchanged_when_care_navigator_not_present_rows = [
+    backdate_job_role_changes_unmapped_codes_rows = [
         ("25", "25"),
         ("40", "40"),
     ]
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/unittest_data/ingest_test_file_data.py:315]: Change the workerid so it doesn't look like a job role code.

File: projects/_01_ingest/ascwds/tests/jobs/test_clean_ascwds_worker_data.py [hunk 5]
```diff
@@ -124,53 +124,48 @@ def test_create_clean_main_job_role_column_returns_expected_main_job_role_clean_
...
+        self.assertEqual(self.returned_data, self.expected_data)
 
-    def test_replace_care_navigator_with_care_coordinator_doesnt_replace_the_value_in_other_columns_in_df(
+    def test_function_doesnt_replace_the_value_in_other_columns_in_df(
         self,
     ):
         self.assertEqual(
             self.returned_data[0][AWKClean.worker_id],
             self.expected_data[0][AWKClean.worker_id],
         )
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/ascwds/tests/jobs/test_clean_ascwds_worker_data.py:156]: Could add in second row of test data here. Checking all the workerid values are unchanged.
        
self.assertEqual(
            self.returned_data[1][AWKClean.worker_id],
            self.expected_data[1][AWKClean.worker_id],
        )

**Review by RoyPriceCS50 — COMMENTED** [review]
Some comments on naming, doc string and test data.

**Review by RoyPriceCS50 — APPROVED** [review]
Looks good.
